From d008ccbdba3ca21f0a93488ca3eacde501e83d51 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Mon, 18 Aug 2014 19:56:20 -0400
Subject: [scsi] qla2xxx: Properly handle 32 bit mailbox register for ISPFX00

Message-id: <1408391817-6389-33-git-send-email-cdupuis@redhat.com>
Patchwork-id: 87922
O-Subject: [PATCH RHEL 7.1 e-stor 32/69] qla2xxx: Properly handle 32 bit mailbox register for ISPFX00.
Bugzilla: 1089346
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Saurav Kashyap <saurav.kashyap@qlogic.com>

scsi commit id 965c77a62b70fe3af3654a92f28d62857e7e6de7

Signed-off-by: Giridhar Malavali <giridhar.malavali@qlogic.com>
Signed-off-by: Saurav Kashyap <saurav.kashyap@qlogic.com>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla2xxx/qla_mr.c b/drivers/scsi/qla2xxx/qla_mr.c
index f3691b8..0b5e510 100644
--- a/drivers/scsi/qla2xxx/qla_mr.c
+++ b/drivers/scsi/qla2xxx/qla_mr.c
@@ -2916,9 +2916,9 @@ qlafx00_async_event(scsi_qla_host_t *vha)
   break;
 
  case QLAFX00_MBA_PORT_UPDATE:  /* Port database update */
-  ha->aenmb[1] = RD_REG_WORD(&reg->aenmailbox1);
-  ha->aenmb[2] = RD_REG_WORD(&reg->aenmailbox2);
-  ha->aenmb[3] = RD_REG_WORD(&reg->aenmailbox3);
+  ha->aenmb[1] = RD_REG_DWORD(&reg->aenmailbox1);
+  ha->aenmb[2] = RD_REG_DWORD(&reg->aenmailbox2);
+  ha->aenmb[3] = RD_REG_DWORD(&reg->aenmailbox3);
   ql_dbg(ql_dbg_async, vha, 0x5077,
       "Asynchronous port Update received "
       "aenmb[0]: %x, aenmb[1]: %x, aenmb[2]: %x, aenmb[3]: %x\n",
@@ -2975,7 +2975,7 @@ static void
 qlafx00_mbx_completion(scsi_qla_host_t *vha, uint32_t mb0)
 {
  uint16_t cnt;
- uint16_t __iomem *wptr;
+ uint32_t __iomem *wptr;
  struct qla_hw_data *ha = vha->hw;
  struct device_reg_fx00 __iomem *reg = &ha->iobase->ispfx00;
 
@@ -2985,10 +2985,10 @@ qlafx00_mbx_completion(scsi_qla_host_t *vha, uint32_t mb0)
  /* Load return mailbox registers. */
  ha->flags.mbox_int = 1;
  ha->mailbox_out32[0] = mb0;
- wptr = (uint16_t __iomem *)&reg->mailbox17;
+ wptr = (uint32_t __iomem *)&reg->mailbox17;
 
  for (cnt = 1; cnt < ha->mbx_count; cnt++) {
-  ha->mailbox_out32[cnt] = RD_REG_WORD(wptr);
+  ha->mailbox_out32[cnt] = RD_REG_DWORD(wptr);
   wptr++;
  }
 }
-- 
1.7.1